dipy/segment/mrf.pyx

```diff
@@ -0,0 +1,729 @@
+#cython: wraparound=False
+#cython: cdivision=True
+import numpy as np
+from dipy.segment.mask import applymask
+from dipy.core.ndindex import ndindex
+from dipy.sims.voxel import add_noise
+cimport cython
+cimport numpy as cnp
+cdef extern from "dpy_math.h" nogil:
+    cdef double NPY_PI
+    cdef double NPY_INFINITY
+    double sqrt(double)
+    double log(double)
+    double exp(double)
+    double fabs(double)
+
+
+class ConstantObservationModel(object):
+    
+    r"""
+    Observation model assuming that the intensity of each class is constant.
+    The model parameters are the means $\mu_{k}$ and variances $\sigma_{k}$ 
+    associated with each tissue class. According to this model, the observed 
+    intensity at voxel $x$ is given by $I(x) = \mu_{k} + \eta_{k}$ where $k$ 
+    is the tissue class of voxel $x$, and $\eta_{k}$ is a Gaussian random
+    variable with zero mean and variance $\sigma_{k}^{2}$. The observation
+    model is responsible of computing the negative log-likelihood of observing
+    any given intensity $z$ at each voxel $x$ assuming the voxel belongs to
+    each class $k$. It also provides a default parameter initialization.
+    """
+
+    def __init__(self):
+        
+        r""" Initializes an instance of the ConstantObservationModel class
+        """
+        pass
+
+
+    def initialize_param_uniform(self, image, nclasses):
+        
+        r""" Initializes the means and variances uniformly
+
+        The means are initialized uniformly along the dynamic range of `image`.
+        The variances are set to 1 for all classes
+
+        Parameters
+        ----------
+        image : array,
+                3D structural image
+        nclasses : int,
+                    number of desired classes
+
+        Returns
+        -------
+        mu : array, 
+                1 x nclasses, mean for each class
+        sigma : array, 
+                    1 x nclasses, standard deviation for each class.
+                    Set up to 1.0 for all classes.
+        """
+
+        cdef:
+            double[:] mu = np.empty((nclasses,), dtype=np.float64)
+            double[:] sigma = np.empty((nclasses,), dtype=np.float64)
+
+        _initialize_param_uniform(image, mu, sigma)
+        return np.array(mu), np.array(sigma)
+
+
+    def seg_stats(self, input_image, seg_image, nclass):
+        
+        r""" Mean and standard variation for N desired  tissue classes
+
+        Parameters
+        ----------
+        input_image : ndarray,
+                        3D structural image
+        seg_image : ndarray,
+                        3D segmented image
+        nclass : int,
+                    number of classes (3 in most cases)
+
+        Returns
+        -------
+        mu, std: ndarrays,
+                    1 x nclasses dimension
+                    Mean and standard deviation for each class
+
+        """
+        mu = np.zeros(nclass)
+        std = np.zeros(nclass)
+
+        for i in range(0, nclass):

```

Comment:
Instead of doing that you can initialize `mu[...]` and `std[...]` to zero. Then iterate over each voxel `(x, y, z)`, grab its label `s = seg_image[x, y, z]` and its value `v = input_image[x, y, z]`. Then update only the entry in `mu`, `std` corresponding to labe `s`: `mu[s] += v`, `std[s] += v*v`. Since you are sweeping the full volume only once and for each voxel you are updating only one entry of the arrays `mu` and `std`, the full complexity is just `Theta(N^3)`.